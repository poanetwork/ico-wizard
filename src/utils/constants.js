export const defaultState = { 
	contracts: { 
		token: {}, 
		crowdsale: {} 
	}, 
	token: {},
	crowdsale: {}
}
