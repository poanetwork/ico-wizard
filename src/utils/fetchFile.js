export async function fetchFile(path) {
  let response = await fetch(path)
  let data = await response.text()
  return data
}
