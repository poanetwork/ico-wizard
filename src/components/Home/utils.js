import { DOWNLOAD_STATUS } from '../../utils/constants'
import { getNetworkVersion } from '../../utils/blockchainHelpers'
import { getCrowdsaleAssets } from '../../stores/utils'
import logdown from 'logdown'

const logger = logdown('TW:home:utils')

export const reloadStorage = async props => {
  let { generalStore, contractStore } = props

  try {
    if (!contractStore || !generalStore) {
      throw new Error('There is no stores to set')
    }
    contractStore.setProperty('downloadStatus', DOWNLOAD_STATUS.PENDING)

    // General store, check network
    let networkID = await getNetworkVersion()
    generalStore.setProperty('networkID', networkID)

    // Contract store, get contract and abi
    await getCrowdsaleAssets(networkID)
    contractStore.setProperty('downloadStatus', DOWNLOAD_STATUS.SUCCESS)
    return true
  } catch (e) {
    logger.error('Error downloading contracts', e)
    if (contractStore) {
      contractStore.setProperty('downloadStatus', DOWNLOAD_STATUS.FAILURE)
    }
    throw e
  }
}
