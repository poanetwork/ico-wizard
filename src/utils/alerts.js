import sweetAlert2 from 'sweetalert2'
import { DEPLOYMENT_VALUES, LIMIT_RESERVED_ADDRESSES, LIMIT_WHITELISTED_ADDRESSES } from './constants'

export const noMetaMaskAlert = cb => {
  sweetAlert2({
    title: 'Warning',
    html:
      "You don't have  <a href='https://chrome.google.com/webstore/detail/nifty-wallet/jbdaocneiiinmjbjlgalhcelgbejmnid'>" +
      "a Wallet </a> installed. Check Token Wizard GitHub for <a href='https://github.com/poanetwork/token-wizard' target='blank'>the instruction</a>.",
    type: 'warning'
  }).then(cb)
}

export const MetaMaskIsLockedAlert = cb => {
  sweetAlert2({
    title: 'Warning',
    html: 'You signed out from your Wallet. Open your Wallet extension and sign in.',
    type: 'warning'
  }).then(cb)
}

export const noContractDataAlert = () => {
  sweetAlert2({
    title: 'Warning',
    html: 'The crowdsale data is empty. There is nothing to deploy. Please, start Token Wizard from the beginning.',
    type: 'warning'
  })
}

export const noContractAlert = () => {
  sweetAlert2({
    title: 'Warning',
    html: 'There is no contract at this address',
    type: 'warning'
  })
}

export const invalidCrowdsaleExecIDAlert = () => {
  sweetAlert2({
    title: 'Warning',
    html: 'Invalid crowdsale exec-id is indicated in config and/or in query string.',
    type: 'warning'
  })
}

export const invalidCrowdsaleExecIDProxyAlert = () => {
  sweetAlert2({
    title: 'Warning',
    html: 'Invalid crowdsale exec-id or proxy address indicated in config and/or in query string.',
    type: 'warning'
  })
}

export const invalidCrowdsaleProxyAlert = () => {
  sweetAlert2({
    title: 'Warning',
    html: 'Invalid crowdsale proxy address is indicated in config and/or in query string.',
    type: 'warning'
  })
}

export const invalidCrowdsaleAddrAlert = () => {
  sweetAlert2({
    title: 'Warning',
    html: 'Invalid crowdsale address is indicated in config and/or in query string.',
    type: 'warning'
  })
}

export const invalidNetworkIDAlert = () => {
  sweetAlert2({
    title: 'Warning',
    html: 'Invalid network ID is indicated in config and/or in query string.',
    type: 'warning'
  })
}

export const successfulContributionAlert = (tokensToContribute, cb) => {
  sweetAlert2({
    title: 'Success',
    html: "Congrats! You've successfully bought " + tokensToContribute + ' tokens!',
    type: 'success'
  }).then(cb)
}

export const contributionDisabledAlertInTime = startTime => {
  sweetAlert2({
    title: 'Warning',
    html: "Wait, please. Crowdsale company hasn't started yet. It'll start from <b>" + new Date(startTime) + '</b>.',
    type: 'warning'
  })
}

export const incorrectNetworkAlert = (correctNetworkName, incorrectNetworkName) => {
  sweetAlert2({
    title: 'Warning',
    html:
      'Crowdsale is from <b>' +
      correctNetworkName +
      ' network</b>. But you are connected to <b>' +
      incorrectNetworkName +
      ' network</b>. Please, change connection in your Wallet extension.',
    type: 'warning'
  })
}

export const warningOnMainnetAlert = (tiersCount, priceSelected, reservedCount, whitelistCount, cb) => {
  const { GAS_REQUIRED, TX_REQUIRED } = DEPLOYMENT_VALUES

  let gasRequired =
    GAS_REQUIRED.DEFAULT + GAS_REQUIRED.WHITELIST * whitelistCount + GAS_REQUIRED.RESERVED_TOKEN * reservedCount

  let txRequired = TX_REQUIRED.DEFAULT
  if (whitelistCount) txRequired += TX_REQUIRED.WHITELIST
  if (reservedCount) txRequired += TX_REQUIRED.RESERVED_TOKEN

  const weiToGwei = x => x / 1000000000
  const n = 100 //fraction to round
  const deployCostInEth = weiToGwei(gasRequired * weiToGwei(priceSelected))
  const estimatedCost = (1.0 / n) * Math.ceil(n * tiersCount * deployCostInEth)
  const estimatedTxsCount = tiersCount * txRequired

  sweetAlert2({
    title: 'Warning',
    html: `You are about to sign ${estimatedTxsCount} TXs. You will see an individual Wallet windows for each of it.
     Please don't open two or more instances of Wizard in one browser. Token Wizard will create ${tiersCount}-tier(s)
     crowdsale for you. The total cost will be around ${estimatedCost.toFixed(2)} ETH. Are you sure you want to
     proceed?`,
    type: 'warning',
    showCancelButton: true,
    confirmButtonText: 'Yes, I am sure!',
    cancelButtonText: 'No, cancel it!'
  }).then(result => {
    if (result.value) {
      cb()
    }
  })
}

export const warningOnFinalizeCrowdsale = () => {
  return sweetAlert2({
    title: 'Finalize Crowdsale',
    html:
      "Are you sure to finalize the crowdsale? After finalization, it's not possible to buy tokens, all tokens will be movable, reserved tokens will be issued. Press <strong>Yes</strong> to finalize, and <strong>NO</strong> to close this dialog.",
    type: 'warning',
    showCancelButton: true,
    cancelButtonText: 'NO',
    confirmButtonText: 'Yes',
    reverseButtons: true
  })
}

export const successfulFinalizeAlert = () => {
  return sweetAlert2({
    title: 'Success',
    html: "Congrats! You've successfully finalized the Crowdsale!",
    type: 'success'
  })
}

export const successfulDistributeAlert = () => {
  sweetAlert2({
    title: 'Success',
    html: "Congrats! You've successfully distributed reserved tokens!",
    type: 'success'
  })
}

export const noGasPriceAvailable = () => {
  sweetAlert2({
    title: 'No Gas Price Available',
    html: "Token Wizard wasn't able to request current Gas Prices from the blockchain, custom values will be used",
    type: 'warning'
  })
}

export const successfulUpdateCrowdsaleAlert = () => {
  sweetAlert2({
    title: 'Success',
    html: "Congrats! You've successfully updated the Crowdsale!",
    type: 'success'
  }).then(result => {
    if (result.value) {
      window.location.reload()
    }
  })
}

export const successfulDeployment = () => {
  sweetAlert2({
    title: 'Success',
    html: 'Transactions signed successfully!',
    type: 'success'
  })
}

export const mainnetIsOnMaintenance = () => {
  sweetAlert2({
    title: 'Warning',
    html:
      "Token Wizard on Mainnet is down for maintenance. For updates, please check <a href='https://gitter.im/poanetwork/token-wizard'>our gitter</a>",
    type: 'warning'
  })
}

export const notTheOwner = () => {
  sweetAlert2({
    title: 'Not The Owner',
    html: "Current user is not the owner of the Crowdsale, thus you won't be able to modify it",
    type: 'warning'
  })
}

export const cancellingIncompleteDeploy = () => {
  return sweetAlert2({
    title: 'Cancel crowdsale deploy',
    html: 'Are you sure you want to cancel the deployment of the crowdsale? This action cannot be undone.',
    type: 'warning',
    showCancelButton: true,
    cancelButtonText: 'No',
    confirmButtonText: 'Yes',
    reverseButtons: true
  })
}

export const skippingTransaction = () => {
  return sweetAlert2({
    title: 'Skip transaction',
    html:
      'Are you sure you want to skip the transaction? This can leave the whole crowdsale in an invalid state, only do this if you are sure of what you are doing.',
    type: 'warning',
    showCancelButton: true,
    cancelButtonText: 'No',
    confirmButtonText: 'Yes',
    reverseButtons: true
  })
}

export const clearingReservedTokens = () => {
  return sweetAlert2({
    title: 'Clear all reserved tokens',
    html: 'Are you sure you want to remove all reserved tokens?',
    type: 'warning',
    showCancelButton: true,
    cancelButtonText: 'No',
    confirmButtonText: 'Yes',
    reverseButtons: true
  })
}

export const clearingWhitelist = () => {
  return sweetAlert2({
    title: 'Clear all whitelist addresses',
    html: 'Are you sure you want to remove all the whitelist addresses for this tier?',
    type: 'warning',
    showCancelButton: true,
    cancelButtonText: 'No',
    confirmButtonText: 'Yes',
    reverseButtons: true
  })
}

export const whitelistImported = count => {
  return sweetAlert2({
    title: 'Addresses imported',
    html: `${count} addresses were added to the whitelist`,
    type: 'info'
  })
}

export const errorDimCSVAlert = dimensions => {
  let message = 'There was an error importing the file.'

  if (dimensions && dimensions.length > 0) {
    const messageHeader = `<div class="text-left">The following dimensions are wrong: </div>`

    let list = '<ul class="text-left">'
    for (let error of dimensions) {
      const { line, dim } = error
      let item = `<li>The line number ${line} has an incorrect dimension: ${dim}</li>`
      list = `${list}${item}`
    }
    list = `${list}</ul>`
    message = `${messageHeader} ${list}`
  }

  return sweetAlert2({
    title: 'Error importing the file',
    html: message,
    type: 'error'
  })
}

export const errorDimValueCSVAlert = dimensionsValues => {
  let message = 'There was an error importing the file.'

  if (dimensionsValues && dimensionsValues.length > 0) {
    const messageHeader = `<div class="text-left">The following value dimensions are wrong: </div>`

    let list = '<ul class="text-left">'
    for (let error of dimensionsValues) {
      const { line, value, validationType } = error
      let item = `<li>The line number ${line} has an incorrect value dimension: ${value}. The field must be ${validationType}.</li>`
      list = `${list}${item}`
    }
    list = `${list}</ul>`
    message = `${messageHeader} ${list}`
  }

  return sweetAlert2({
    title: 'Error importing the file',
    html: message,
    type: 'error'
  })
}

export const errorEmptyCSVAlert = () => {
  return sweetAlert2({
    title: 'Error importing the file',
    html: 'Empty CSV file. Nothing was imported',
    type: 'error'
  })
}

export const errorWhitelistedCSVAlert = errors => {
  let message = 'There was an error importing the file. The file have an erroneous amount of columns'

  if (errors && errors.length > 0) {
    const plural1 = errors.length > 1 ? 'lines' : 'line'
    const plural2 = errors.length > 1 ? 'are' : 'is'
    const messageHeader = `<div class="text-left">The following ${plural1} ${plural2} wrong: </div>`

    let list = `<ul class="text-left"><li>${errors.join(`</li><li>`)}</ul>`
    message = `${messageHeader} ${list}`
  }

  return sweetAlert2({
    title: 'Error importing the file',
    html: message,
    type: 'error'
  })
}

export const errorRowLengthCSVAlert = errors => {
  let message = 'There was an error importing the file. The file have an erroneous amount of columns'

  if (errors && errors.length > 0) {
    const messageHeader = `<div class="text-left">The following lines have an erroneous amount of columns: </div>`

    let list = '<ul class="text-left">'
    for (let error of errors) {
      const { line, columns } = error
      const columnLabel = columns > 1 ? 'columns' : 'column'
      let item = `<li>The line number ${line} have ${columns} ${columnLabel}, must have 3 columns</li>`
      list = `${list}${item}`
    }
    list = `${list}</ul>`
    message = `${messageHeader} ${list}`
  }

  return sweetAlert2({
    title: 'Error importing the file',
    html: message,
    type: 'error'
  })
}

export const errorAddressCSVAlert = address => {
  let message = 'There was an error importing the file.'

  if (address && address.length > 0) {
    const messageHeader = `<div class="text-left">The following address are wrong: </div>`

    let list = '<ul class="text-left">'
    for (let error of address) {
      const { line, address } = error
      let item = `<li>The line number ${line} has an incorrect address: ${address}</li>`
      list = `${list}${item}`
    }
    list = `${list}</ul>`
    message = `${messageHeader} ${list}`
  }

  return sweetAlert2({
    title: 'Error importing the file',
    html: message,
    type: 'error'
  })
}

export const reservedTokensImported = count => {
  return sweetAlert2({
    title: 'Reserved tokens imported',
    html: `Tokens will be reserved for ${count} addresses`,
    type: 'info'
  })
}

export const noMoreTokensAvailable = () => {
  return sweetAlert2({
    title: 'No more tokens available',
    html: `You're not able to buy more tokens. You have bought your maximum allowed or tier has reached its hard cap`,
    type: 'info'
  })
}

export const notAllowedContributor = () => {
  return sweetAlert2({
    title: 'Not allowed',
    html: `You're not allowed to buy tokens during this tier.`,
    type: 'info'
  })
}

export const noMoreReservedSlotAvailable = () => {
  return sweetAlert2({
    title: 'No more reserved tokens available',
    html: `You're not able to reserve more tokens. The maximum allowed is ${LIMIT_RESERVED_ADDRESSES}`,
    type: 'info'
  })
}

export const noMoreReservedSlotAvailableCSV = count => {
  return sweetAlert2({
    title: 'You reach the limit of reserved tokens',
    html: `You're not able to reserve more tokens. Reserved tokens imported Tokens will be reserved for ${count} addresses. The maximum allowed is ${LIMIT_RESERVED_ADDRESSES}`,
    type: 'info'
  })
}

export const noMoreWhitelistedSlotAvailable = () => {
  return sweetAlert2({
    title: 'Maximum limit of addresses reached',
    html: `You're not able to add more addresses to the whitelist. The maximum allowed is ${LIMIT_WHITELISTED_ADDRESSES}`,
    type: 'info'
  })
}

export const noMoreWhitelistedSlotAvailableCSV = count => {
  return sweetAlert2({
    title: 'Maximum limit of addresses reached',
    html: `You're not able to add more addresses to the whitelist. Only ${count} addresses of the file could be added. The maximum allowed is ${LIMIT_WHITELISTED_ADDRESSES}`,
    type: 'info'
  })
}

export const maxCapExceedsSupply = count => {
  return sweetAlert2({
    title: `Max Cap exceeded tier's supply`,
    html: `<strong>${count}</strong> addresses where added. Addresses with <code>maxCap</code> greater than tier's supply where discarded.`,
    type: 'info'
  })
}

export const networkChanged = () => {
  return sweetAlert2({
    title: 'Network changed',
    html: 'The network id was modified, please set up nifty wallet with the network id which you started',
    type: 'warning'
  })
}

export const deployHasEnded = () => {
  return sweetAlert2({
    title: 'Deploy ended',
    html: 'The deploy is finished',
    type: 'info'
  })
}

export const transactionLost = () => {
  return sweetAlert2({
    title: 'Tx Lost',
    html: "Please cancel pending transaction, if there's any, in your wallet (Nifty Wallet or Metamask) and Continue",
    type: 'error',
    confirmButtonText: 'Continue'
  })
}
