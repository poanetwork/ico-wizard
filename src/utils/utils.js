import { TOAST } from './constants'
import queryString from 'query-string'
import { BigNumber } from 'bignumber.js'
import logdown from 'logdown'
import Web3 from 'web3'
import moment from 'moment'
import { fetchFile } from './fetchFile'
import { isObservableArray } from 'mobx'

const logger = logdown('TW:utils:utils')

export const getQueryVariable = variable => {
  return queryString.parse(window.location.search)[variable]
}

export const isExecIDValid = execID => /^0x[a-f0-9]{64}$/i.test(execID)

export const isAddressValid = address => Web3.utils.isAddress(address)

export const getExecIDFromQuery = () => {
  const execID = getQueryVariableExecId()
  logger.log('getExecID:', execID)
  return isExecIDValid(execID) ? execID : null
}

export const getAddrFromQuery = () => {
  const addr = getQueryVariableAddr()
  logger.log('getAddr:', addr)
  return isAddressValid(addr) ? addr : null
}

export const getQueryVariableAddr = () => {
  return getQueryVariable('addr')
}

export const getQueryVariableExecId = () => {
  return getQueryVariable('exec-id')
}

export const isNetworkIDValid = networkID => /^[0-9]+$/.test(networkID)

export const getNetworkID = () => {
  const networkID = getQueryVariable('networkID')
  return isNetworkIDValid(networkID) ? networkID : null
}

export const setFlatFileContentToState = async file => {
  return await fetchFile(file)
}

export const dateToTimestamp = date => new Date(date).getTime()

export const validateTier = tier => typeof tier === 'string' && tier.length > 0 && tier.length < 30

export const validateSupply = supply => isNaN(Number(supply)) === false && Number(supply) > 0

export const validateTime = time => dateToTimestamp(time) > Date.now()

export const validateLaterTime = (laterTime, previousTime) => dateToTimestamp(laterTime) > dateToTimestamp(previousTime)

export const validateLaterOrEqualTime = (laterTime, previousTime) =>
  dateToTimestamp(laterTime) >= dateToTimestamp(previousTime)

export const toFixed = x => {
  if (Math.abs(x) < 1.0) {
    let e = parseInt(x.toString().split('e-')[1], 10)
    if (e) {
      x *= Math.pow(10, e - 1)
      x = '0.' + new Array(e).join('0') + x.toString().substring(2)
    }
  } else {
    let e = parseInt(x.toString().split('+')[1], 10)
    if (e > 20) {
      e -= 20
      x /= Math.pow(10, e)
      x += new Array(e + 1).join('0')
    }
  }
  return x
}

export const toast = {
  msg: {},
  showToaster: function({ type = TOAST.TYPE.INFO, message = '', options = {} }) {
    if (!message) {
      return
    }

    if (typeof this.msg[type] === 'function') {
      this.msg[type](message, options)
    } else {
      return
    }
  }
}

export const gweiToWei = x => parseInt(x * 1000000000, 10)

export const weiToGwei = x => x / 1000000000

export const countDecimalPlaces = num => {
  /*
    (?:
      \.
      (\d+)  First captured group: decimals after the point but before the e
    )?
    (?:
      [eE]
      ([+-]?\d+)  Second captured group: exponent used to adjust the count
    )?
    $
  */
  const match = ('' + num).match(/(?:\.(\d+))?(?:[eE]([+-]?\d+))?$/)

  if (!match[0] && !match[1] && !match[2]) return 0

  const digitsAfterDecimal = match[1] ? match[1].length : 0
  const adjust = match[2] ? +match[2] : 0

  return Math.max(0, digitsAfterDecimal - adjust)
}

export const acceptPositiveIntegerOnly = value => {
  if (typeof value === 'number') value = String(value)
  if (typeof value !== 'string') return ''

  return String(value).match(/^(\d*)/)[1]
}

export const removeTrailingNUL = ascii => ascii.replace(/\x00+$/, '')

export const truncateStringInTheMiddle = (str, strLength = 50, strPositionStart = 24, strPositionEnd = 25) => {
  if (typeof str === 'string' && str.length > strLength) {
    return `${str.substr(0, strPositionStart)}...${str.substr(str.length - strPositionEnd, str.length)}`
  }
  return str
}

/**
 * Converts the value passed to a BigNumber instance
 * @param {*} value - A number representation
 * @param {boolean} [force=true] - If set to false will return 'undefined' when value is not a number or a string
 * representation of a number.
 * @returns {BigNumber|undefined}
 */
export const toBigNumber = (value, force = true) => {
  BigNumber.set({ DECIMAL_PLACES: 18 })

  if (isNaN(value) || value === '' || value === null) {
    if (force) {
      return new BigNumber(0)
    } else {
      return undefined
    }
  } else {
    return new BigNumber(value)
  }
}

/**
 * Sleep function like C
 * @param ms
 * @returns {Promise}
 */
export const sleep = async ms => {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export const clearStorage = props => {
  // Generate of stores to clear
  const toArray = ({
    generalStore,
    contractStore,
    crowdsaleStore,
    gasPriceStore,
    deploymentStore,
    reservedTokenStore,
    stepTwoValidationStore,
    tierStore,
    tokenStore
  }) => {
    return [
      generalStore,
      contractStore,
      crowdsaleStore,
      gasPriceStore,
      deploymentStore,
      reservedTokenStore,
      stepTwoValidationStore,
      tierStore,
      tokenStore
    ]
  }

  const storesToClear = toArray(props)
  for (let storeToClear of storesToClear) {
    if (storeToClear && typeof storeToClear.reset === 'function') {
      logger.log('Store to be cleared:', storeToClear.constructor.name)
      storeToClear.reset()
    }
  }
}

export const navigateTo = data => {
  const { history, location, params = '', fromLocation } = data
  const path = convertLocationToPath(location)

  if (path === null) {
    throw new Error(`invalid location specified: ${location}`)
  }

  if (!history || !(typeof history === 'object')) {
    throw new Error(`invalid history object: ${history}`)
  }

  history.push({
    pathname: `${path}${params}`,
    state: {
      fromLocation: fromLocation
    }
  })

  return true
}

export const convertLocationToPath = location => {
  return (
    {
      home: '/',
      stepOne: '1',
      stepTwo: '2',
      stepThree: '3',
      stepFour: '4',
      manage: 'manage',
      crowdsales: 'crowdsales',
      contribute: 'contribute'
    }[location] || null
  )
}

export const goBack = data => {
  const { history } = data

  if (!history || !(typeof history === 'object')) {
    throw new Error(`invalid history object: ${history}`)
  }

  if (typeof history.goBack === undefined) {
    throw new Error(`invalid goBack function`)
  }

  history.goBack()

  return true
}

export const goBackMustBeEnabled = data => {
  const { history } = data

  if (!history || !(typeof history === 'object')) {
    throw new Error(`invalid history object: ${history}`)
  }
  const length = history.length || 0
  return length > 1
}

export const convertDateToLocalTimezoneInUnix = dateToConvert => {
  const dateConvertUnix = moment(dateToConvert)
    .local()
    .unix()
  return dateConvertUnix * 1000
}

export const convertDateToUTCTimezoneToDisplay = dateToConvert => {
  return moment(dateToConvert)
    .utc()
    .format('YYYY-MM-DD HH:mm (z ZZ)')
}

export const getContractBySourceType = (sourceType, isMintedCappedCrowdsale, { DutchProxy, MintedCappedProxy }) => {
  const parseContent = content => (isObservableArray(content) ? JSON.stringify(content.slice()) : content)
  return isMintedCappedCrowdsale ? parseContent(MintedCappedProxy[sourceType]) : parseContent(DutchProxy[sourceType])
}

export const getSourceTypeTitle = sourceType => {
  const sourceTypeName = {
    abi: 'ABI',
    bin: 'Creation Code',
    src: 'Source Code'
  }

  return `Crowdsale Proxy Contract ${sourceTypeName[sourceType]}`
}

export const updateProxyContractInfo = ({ contractAddress }, { web3Store, contractStore, crowdsaleStore }, params) => {
  const { web3 } = web3Store
  const encoded = web3.eth.abi.encodeParameters(['address', 'bytes32', 'address', 'bytes32'], params)

  contractStore.setContractProperty(crowdsaleStore.proxyName, 'addr', contractAddress.toLowerCase())
  contractStore.setContractProperty(crowdsaleStore.proxyName, 'abiEncoded', encoded.slice(2))
}

export const downloadFile = (data, filename, mimetype) => {
  if (!data) {
    return
  }

  const blob = data.constructor !== Blob ? new Blob([data], { type: mimetype || 'application/octet-stream' }) : data

  if (navigator.msSaveBlob) {
    navigator.msSaveBlob(blob, filename)
    return
  }

  const lnk = document.createElement('a')
  const url = window.URL
  let objectURL

  if (mimetype) {
    lnk.type = mimetype
  }

  lnk.download = filename || 'untitled'
  lnk.href = objectURL = url.createObjectURL(blob)
  lnk.dispatchEvent(new MouseEvent('click'))
  setTimeout(url.revokeObjectURL.bind(url, objectURL))
}

export const uniqueElementsBy = (arr, fn) =>
  arr.reduce((acc, v) => {
    if (!acc.some(x => fn(v, x))) acc.push(v)
    return acc
  }, [])

export const pad = (num, size) => {
  let s = num + ''
  while (s.length < size) s = '0' + s
  return s
}
