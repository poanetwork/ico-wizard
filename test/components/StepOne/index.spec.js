import React from 'react'
import Adapter from 'enzyme-adapter-react-15'
import { configure, mount, shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router'
import { StepOne } from '../../../src/components/StepOne/index'
import {
  web3Store,
  generalStore,
  contractStore,
  crowdsaleStore,
  gasPriceStore,
  deploymentStore,
  reservedTokenStore,
  stepTwoValidationStore,
  tierStore,
  tokenStore
} from '../../../src/stores'
import { CROWDSALE_STRATEGIES } from '../../../src/utils/constants'
import GasPriceInput from '../../../src/components/StepThree/GasPriceInput'

configure({ adapter: new Adapter() })

describe('StepOne', () => {
  const history = { push: jest.fn() }
  const stores = {
    web3Store,
    generalStore,
    contractStore,
    crowdsaleStore,
    gasPriceStore,
    deploymentStore,
    reservedTokenStore,
    stepTwoValidationStore,
    tierStore,
    tokenStore
  }
  const { MINTED_CAPPED_CROWDSALE, DUTCH_AUCTION } = CROWDSALE_STRATEGIES

  it(`should render StepOne screen`, () => {
    // Given
    const component = renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <StepOne {...stores} />
      </MemoryRouter>
    )

    // When
    const tree = component.toJSON()

    // Then
    expect(tree).toMatchSnapshot()
  })

  it(`should navigate to StepTwo`, () => {
    // Given
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']}>
        <StepOne {...stores} history={history} />
      </MemoryRouter>
    )
    const stepOneComponent = wrapper.find('StepOne')
    const navigateToHandler = jest.spyOn(stepOneComponent.instance(), 'goNextStep')
    wrapper.update()

    // When
    stepOneComponent.find('.sw-ButtonContinue').simulate('click')

    // Then
    expect(navigateToHandler).toHaveBeenCalledTimes(1)
    expect(navigateToHandler).toHaveBeenCalledWith()
  })

  it(`should switch to DutchAuction Strategy`, () => {
    // Given
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']}>
        <StepOne {...stores} history={history} />
      </MemoryRouter>
    )
    const stepOneComponent = wrapper.find('StepOne')

    // When
    stepOneComponent
      .find('input[name="contract-type"]')
      .at(1)
      .simulate('change', { target: { value: DUTCH_AUCTION } })

    // Then
    expect(stepOneComponent.instance().state.strategy).toBe(DUTCH_AUCTION)
  })

  it(`should switch to MintedCapped Strategy`, () => {
    // Given
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']}>
        <StepOne {...stores} history={history} />
      </MemoryRouter>
    )
    const stepOneComponent = wrapper.find('StepOne')

    // When
    const contractTypeInputs = stepOneComponent.find('input[name="contract-type"]')
    contractTypeInputs.at(1).simulate('change', { target: { value: DUTCH_AUCTION } })
    contractTypeInputs.at(0).simulate('change', { target: { value: MINTED_CAPPED_CROWDSALE } })

    // Then
    expect(stepOneComponent.instance().state.strategy).toBe(MINTED_CAPPED_CROWDSALE)
  })

  it(`should render StepOne screen and test load method`, async () => {
    // Given
    const wrapper = shallow(<StepOne {...stores} />)
    // When
    const result = await wrapper
      .dive()
      .instance()
      .load()
    // Then
    expect(result).toEqual({ strategy: 'white-list-with-cap' })
  })

  it(`should render StepOne screen and test load method with clearStorage`, async () => {
    // Given
    global.localStorage.clearStorage = true
    const wrapper = shallow(<StepOne {...stores} />)
    // When
    const result = await wrapper
      .dive()
      .instance()
      .load()
    // Then
    setTimeout(() => {
      expect(result).toEqual({ strategy: 'white-list-with-cap' })
    }, 2000)
  })

  it(`should render StepOne screen and test reload`, async () => {
    // Given
    global.localStorage.reload = true
    global.location = jest.fn()
    global.location.assign = jest.fn()
    const wrapper = mount(<StepOne {...stores} />)

    // When
    setTimeout(() => {
      // Then
      expect(global.localStorage.reload).toBe(undefined)
      expect(global.localStorage.clearStorage).toBeTruthy()
    }, 2000)
  })

  it(`should render StepOne screen and test beforeUnloadSpy`, async () => {
    const wrapper = mount(<StepOne {...stores} />)
    window.location.reload()
    setTimeout(() => {
      expect(global.localStorage.reload).toBe(undefined)
    }, 2000)
  })
})
