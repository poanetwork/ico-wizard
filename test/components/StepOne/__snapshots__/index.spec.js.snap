// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepOne should render StepOne screen 1`] = `
<div>
  <section
    className="lo-MenuBarAndContent"
  >
    <div
      className="st-StepNavigation"
    >
      <a
        className="sw-LogoWhite"
        href="/"
      />
      <div
        className="st-StepNavigation_Container"
      >
        <div
          className="st-StepItem
        active  "
        >
          <div
            className="st-StepItem_Circle"
          />
          <div
            className="st-StepItem_Line"
          />
          <span
            className="st-StepItem_Text"
          >
            Crowdsale Strategy
          </span>
        </div>
        <div
          className="st-StepItem
          "
        >
          <div
            className="st-StepItem_Circle"
          />
          <div
            className="st-StepItem_Line"
          />
          <span
            className="st-StepItem_Text"
          >
            Token Setup
          </span>
        </div>
        <div
          className="st-StepItem
          "
        >
          <div
            className="st-StepItem_Circle"
          />
          <div
            className="st-StepItem_Line"
          />
          <span
            className="st-StepItem_Text"
          >
            Crowdsale Setup
          </span>
        </div>
        <div
          className="st-StepItem
          "
        >
          <div
            className="st-StepItem_Circle"
          />
          <div
            className="st-StepItem_Line"
          />
          <span
            className="st-StepItem_Text"
          >
            Publish
          </span>
        </div>
        <div
          className="st-StepItem
          "
        >
          <div
            className="st-StepItem_Circle"
          />
          <span
            className="st-StepItem_Text"
          >
            Crowdsale Page
          </span>
        </div>
      </div>
    </div>
    <div
      className="st-StepContent"
    >
      <div
        className="st-SectionInfo "
      >
        <div
          className="st-SectionInfo_Icon st-SectionInfo_Icon-step1"
        />
        <div
          className="st-SectionInfo_Text"
        >
          <h1
            className="st-SectionInfo_Title"
          >
            Crowdsale Strategy
          </h1>
          <p
            className="st-SectionInfo_Description"
            dangerouslySetInnerHTML={
              Object {
                "__html": "Select a strategy for your crowdsale contract.",
              }
            }
          />
        </div>
      </div>
      <div
        className="sw-RadioItems"
      >
        <label
          className="st-StrategyItem_Item"
        >
          <input
            checked={true}
            className="st-StrategyItem_InputRadio"
            id="white-list-with-cap"
            name="contract-type"
            onChange={[Function]}
            type="radio"
            value="white-list-with-cap"
          />
          <span
            className="st-StrategyItem_ItemContent"
          >
            <span
              className="st-StrategyItem_ItemContentText"
            >
              <span
                className="st-StrategyItem_ItemTitle"
              >
                Whitelist with Cap
              </span>
              <span
                className="st-StrategyItem_ItemDescription"
              >
                Modern crowdsale strategy with multiple tiers, whitelists, and limits. Recommended for every crowdsale.
              </span>
            </span>
            <span
              className="st-StrategyItem_Radio"
            />
          </span>
        </label>
        <label
          className="st-StrategyItem_Item"
        >
          <input
            checked={false}
            className="st-StrategyItem_InputRadio"
            id="dutch-auction"
            name="contract-type"
            onChange={[Function]}
            type="radio"
            value="dutch-auction"
          />
          <span
            className="st-StrategyItem_ItemContent"
          >
            <span
              className="st-StrategyItem_ItemContentText"
            >
              <span
                className="st-StrategyItem_ItemTitle"
              >
                Dutch Auction
              </span>
              <span
                className="st-StrategyItem_ItemDescription"
              >
                An auction with descending price.
              </span>
            </span>
            <span
              className="st-StrategyItem_Radio"
            />
          </span>
        </label>
      </div>
      <div
        className="st-StepContent_Buttons"
      >
        <button
          className="sw-ButtonContinue "
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span
            className="sw-ButtonContinue_Text"
          >
            Continue
          </span>
        </button>
      </div>
    </div>
  </section>
</div>
`;
